restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -219,15 +389,191 @@ private Accumulator combine(Accumulator buffer) {
+      // choose to resume fetching from the next row in current page even if the page itself does
+      // not have a paging state. Therefore, in this case we cannot return EXHAUSTED_PAGE_STATE.
+      if (lastInPage() && page().resultSet().getPagingState() == null) {
+        return EXHAUSTED_PAGE_STATE;
+      }
+
+      return page()
+          .resultSet()
+          .makePagingState(PagingPosition.ofCurrentRow(row()).resumeFrom(resumeMode).build());
+    }
+
+    @Value.Lazy
+    @Override
+    public String toString() {
+      return row().toString();
+    }
+  }
+
+  /** A thin wrapper around {@link ResultSet} that caches {@link #decorator()} objects. */
+  @Value.Immutable(lazyhash = true)
+  public abstract static class Page {
+
+    abstract ResultSet resultSet();
+
+    /** Lazily initialized and cached {@link RowDecorator}. */
+    @Value.Lazy
+    RowDecorator decorator() {
+      return resultSet().makeRowDecorator();
+    }
+  }
+
+  /** A helper class for dealing with a list of related queries. */
+  @Value.Immutable(lazyhash = true)
+  public abstract static class QueryData {
+
+    abstract List<BoundQuery> queries();
+
+    @Value.Lazy
+    List<BuiltSelect> selectQueries() {
+      return queries().stream()
+          .map(q -> (BuiltSelect) q.source().query())
+          .collect(Collectors.toList());
+    }
+
+    @Value.Lazy
+    Table table() {
+      Set<Table> tables =
+          selectQueries().stream().map(BuiltSelect::table).collect(Collectors.toSet());
+
+      if (tables.isEmpty()) {
+        throw new IllegalArgumentException("No tables are referenced by the provided queries");
+      }
+
+      if (tables.size() > 1) {
+        throw new IllegalArgumentException(
+            "Too many tables are referenced by the provided queries: "
+                + tables.stream().map(Table::name).collect(Collectors.joining(", ")));
+      }
+
+      return tables.iterator().next();
+    }
+
+    @Value.Lazy
+    Set<Column> selectedColumns() {
+      Set<Set<Column>> sets =
+          selectQueries().stream().map(BuiltSelect::selectedColumns).collect(Collectors.toSet());
+
+      if (sets.size() != 1) {
+        throw new IllegalArgumentException(
+            "Incompatible sets of columns are selected by the provided queries: "
+                + sets.stream()
+                    .map(
+                        s ->
+                            s.stream().map(Column::name).collect(Collectors.joining(",", "[", "]")))
+                    .collect(Collectors.joining("; ", "[", "]")));
+      }
+
+      return sets.iterator().next();
+    }
+
+    private boolean isSelected(Column column) {
+      // An empty selection set means `*` (all columns)
+      return selectedColumns().isEmpty() || selectedColumns().contains(column);
+    }
+
+    /**
+     * Retrurn the list of columns whose values are used to distinguish one document from another.
+     */
+    private List<Column> docIdColumns(int keyDepth) {
```

Comment:
could this be `@Value.Lazy` as well?

Reply:
sure, but I thought `@Value.Lazy` was not required in this case since this method is not a "getter" (it has an arg).